src/components/RenderTime.tsx

```diff
@@ -1,30 +1,109 @@
 import { css } from '@emotion/css';
-import { PanelProps } from '@grafana/data';
+import { Field, PanelData, PanelProps } from '@grafana/data';
 import moment, { Moment } from 'moment-timezone';
 import React, { useMemo } from 'react';
-import { ClockMode, ClockOptions, ClockType, TimeSettings } from 'types';
+import {
+  ClockMode,
+  ClockOptions,
+  ClockSource,
+  ClockType,
+  CountdownQueryCalculation,
+  CountupQueryCalculation,
+  TimeSettings,
+} from 'types';
 import { getMoment } from 'utils';
 
 function getCountdownText({
   countdownSettings,
   timezone,
+  data,
   replaceVariables,
   now,
 }: {
   countdownSettings: ClockOptions['countdownSettings'];
   timezone: ClockOptions['timezone'];
+  data: PanelData;
   replaceVariables: PanelProps['replaceVariables'];
   now: Moment;
 }): string {
-  if (!countdownSettings.endCountdownTime) {
-    return countdownSettings.endText;
+  let timeLeft = moment.duration(0);
+
+  switch (countdownSettings.source) {
+    case ClockSource.input:
+      if (!countdownSettings.endCountdownTime) {
+        return countdownSettings.endText;
+      }
+
+      timeLeft = moment.duration(
+        moment(replaceVariables(countdownSettings.endCountdownTime))
+          .utcOffset(getMoment(timezone).format('Z'), true)
+          .diff(now)
+      );
+      break;
+    case ClockSource.query:
+      if (data.state === 'Done' && data.series.length > 0) {
+        let field_name = countdownSettings.queryField;
+        if (
```

Comment:
nit: you can reduce all this condition to `!field_name?.length || !data?.series?.[0]?.fields?.length`